core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketsort/BucketSortPipelineAggregator.java

```diff
@@ -0,0 +1,194 @@
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.pipeline.bucketsort;
+
+
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.InternalAggregation.ReduceContext;
+import org.elasticsearch.search.aggregations.InternalMultiBucketAggregation;
+import org.elasticsearch.search.aggregations.bucket.MultiBucketsAggregation;
+import org.elasticsearch.search.aggregations.pipeline.BucketHelpers;
+import org.elasticsearch.search.aggregations.pipeline.BucketHelpers.GapPolicy;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.sort.FieldSortBuilder;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+public class BucketSortPipelineAggregator extends PipelineAggregator {
+
+    private final List<FieldSortBuilder> sorts;
+    private final int from;
+    private final Integer size;
+    private final GapPolicy gapPolicy;
+
+    public BucketSortPipelineAggregator(String name, List<FieldSortBuilder> sorts, int from, Integer size, GapPolicy gapPolicy,
+                                        Map<String, Object> metadata) {
+        super(name, sorts.stream().map(s -> s.getFieldName()).toArray(String[]::new), metadata);
+        this.sorts = sorts;
+        this.from = from;
+        this.size = size;
+        this.gapPolicy = gapPolicy;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public BucketSortPipelineAggregator(StreamInput in) throws IOException {
+        super(in);
+        sorts = in.readList(FieldSortBuilder::new);
+        from = in.readVInt();
+        size = in.readOptionalVInt();
+        gapPolicy = GapPolicy.readFrom(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeList(sorts);
+        out.writeVInt(from);
+        out.writeOptionalVInt(size);
+        gapPolicy.writeTo(out);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return BucketSortPipelineAggregationBuilder.NAME;
+    }
+
+    @Override
+    public InternalAggregation reduce(InternalAggregation aggregation, ReduceContext reduceContext) {
+        InternalMultiBucketAggregation<InternalMultiBucketAggregation, InternalMultiBucketAggregation.InternalBucket> originalAgg =
+                (InternalMultiBucketAggregation<InternalMultiBucketAggregation, InternalMultiBucketAggregation.InternalBucket>) aggregation;
+        List<? extends InternalMultiBucketAggregation.InternalBucket> buckets = originalAgg.getBuckets();
+        int bucketsCount = buckets.size();
+        int offset = reduceContext.isFinalReduce() ? from : 0;
```

Comment:
Pipelines are only called on the final reduce, so this check isn't needed.  I was a bit confused about it too (the javadocs for `isFinalReduce()` is sorta misleading) and had to bug Colin :)  

You can see that the incremental reduction passes `null` for pipelines in `SearchPhaseController`: https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java#L518

So I don't think this check is needed.

Reply:
Ah, fair enough! Will remove.